File: sdk/go/pkg/artifact/language/rust.go
package language

import (
	"bytes"
	"fmt"
	"os"
	"slices"
	"strings"
	"text/template"

	api "github.com/ALT-F4-LLC/vorpal/sdk/go/pkg/api/artifact"
	"github.com/ALT-F4-LLC/vorpal/sdk/go/pkg/artifact"
	"github.com/ALT-F4-LLC/vorpal/sdk/go/pkg/config"
	"github.com/BurntSushi/toml"
)

type RustArtifactCargoToml struct {
	Bin       []RustArtifactCargoTomlBinary   `toml:"bin,omitempty"`
	Package   *RustArtifactCargoTomlPackage   `toml:"package,omitempty"`
	Workspace *RustArtifactCargoTomlWorkspace `toml:"workspace,omitempty"`
}

type RustArtifactCargoTomlBinary struct {
	Name string `toml:"name"`
	Path string `toml:"path"`
}

type RustArtifactCargoTomlPackage struct {
	Name string `toml:"name"`
}

type RustArtifactCargoTomlWorkspace struct {
	Members []string `toml:"members,omitempty"`
}

type Rust struct {
	artifacts []*string
	bins      []string
	build     bool
	check     bool
	excludes  []string
	format    bool
	includes  []string
	lint      bool
	name      string
	packages  []string
	secrets   []*api.ArtifactStepSecret
	source    *string
	tests     bool
	systems   []api.ArtifactSystem
}

type VendorStepScriptTemplateArgs struct {
	Name        string
	Packages    string
	TargetPaths string
}

type StepScriptTemplateArgs struct {
	BinNames      string
	Build         string
	Check         string
	Format        string
	Lint          string
	ManifestPaths string
	Name          string
	Packages      string
	Tests         string
	Vendor        string
}

const VendorStepScriptTemplate = `
mkdir -pv $HOME

pushd ./source/{{.Name}}-vendor
{{if .Packages}}
cat > Cargo.toml << "EOF"
[workspace]
members = [{{.Packages}}]
resolver = "2"
EOF

target_paths=({{.TargetPaths}})

for target_path in ${{"{"}}target_paths{{"["}}@{{"]"}}{{"}"}}; do
    mkdir -pv $(dirname ${{"{"}}target_path{{"}"}})
    touch ${{"{"}}target_path{{"}"}}
done
{{else}}
mkdir -pv src
touch src/main.rs
{{end}}
mkdir -pv $VORPAL_OUTPUT/vendor

cargo_vendor=$(cargo vendor --versioned-dirs $VORPAL_OUTPUT/vendor)

echo "$cargo_vendor" > $VORPAL_OUTPUT/config.toml`

const StepScriptTemplate = `
mkdir -pv $HOME

pushd ./source/{{.Name}}

mkdir -pv .cargo
mkdir -pv $VORPAL_OUTPUT/bin

ln -sv {{.Vendor}}/config.toml .cargo/config.toml
{{if .Packages}}
cat > Cargo.toml << "EOF"
[workspace]
members = [{{.Packages}}]
resolver = "2"
EOF
{{end}}
bin_names=({{.BinNames}})
manifest_paths=({{.ManifestPaths}})

if [ "{{.Format}}" = "true" ]; then
    echo "Running formatter..."
    cargo --offline fmt --all --check
fi

for manifest_path in ${{"{"}}manifest_paths{{"["}}@{{"]"}}{{"}"}}; do
    if [ "{{.Lint}}" = "true" ]; then
        echo "Running linter..."
        cargo --offline clippy --manifest-path ${{"{"}}manifest_path{{"}"}} -- --deny warnings
    fi
done

for bin_name in ${{"{"}}bin_names{{"["}}@{{"]"}}{{"}"}}; do
    if [ "{{.Check}}" = "true" ]; then
        echo "Running check..."
        cargo --offline check --bin ${{"{"}}bin_name{{"}"}} --release
    fi

    if [ "{{.Build}}" = "true" ]; then
        echo "Running build..."
        cargo --offline build --bin ${{"{"}}bin_name{{"}"}} --release
    fi

    if [ "{{.Tests}}" = "true" ]; then
        echo "Running tests..."
        cargo --offline test --bin ${{"{"}}bin_name{{"}"}} --release
    fi

    cp -pv ./target/release/${{"{"}}bin_name{{"}"}} $VORPAL_OUTPUT/bin/
done`

func stripPrefix(path, prefix string) string {
	if strings.HasPrefix(path, prefix) {
		return path[len(prefix)+1:]
	}

	return path
}

func NewRust(name string, systems []api.ArtifactSystem) *Rust {
	return &Rust{
		artifacts: make([]*string, 0),
		bins:      make([]string, 0),
		build:     true,
		check:     false,
		excludes:  make([]string, 0),
		format:    false,
		includes:  make([]string, 0),
		lint:      false,
		name:      name,
		packages:  make([]string, 0),
		secrets:   make([]*api.ArtifactStepSecret, 0),
		source:    nil,
		tests:     false,
		systems:   systems,
	}
}

func (a *Rust) WithArtifacts(artifacts []*string) *Rust {
	a.artifacts = artifacts
	return a
}

func (a *Rust) WithBins(bins []string) *Rust {
	a.bins = bins
	return a
}

func (a *Rust) WithCheck() *Rust {
	a.check = true
	return a
}

func (a *Rust) WithExcludes(excludes []string) *Rust {
	a.excludes = excludes
	return a
}

func (a *Rust) WithFormat() *Rust {
	a.format = true
	return a
}

func (a *Rust) WithIncludes(includes []string) *Rust {
	a.includes = includes
	return a
}

func (a *Rust) WithLint() *Rust {
	a.lint = true
	return a
}

func (a *Rust) WithPackages(packages []string) *Rust {
	a.packages = packages
	return a
}

func (a *Rust) WithSecrets(secrets map[string]string) *Rust {
	for name, value := range secrets {
		secret := &api.ArtifactStepSecret{
			Name:  name,
			Value: value,
		}

		if slices.ContainsFunc(a.secrets, func(s *api.ArtifactStepSecret) bool { return s.Name == name }) {
			continue
		}

		a.secrets = append(a.secrets, secret)
	}

	return a
}

func (a *Rust) WithSource(source *string) *Rust {
	a.source = source
	return a
}

func (a *Rust) WithTests() *Rust {
	a.tests = true
	return a
}

func (builder *Rust) Build(context *config.ConfigContext) (*string, error) {
	protoc, err := artifact.Protoc(context)
	if err != nil {
		return nil, err
	}

	// Parse source path

	contextPath := context.GetArtifactContextPath()

	sourcePath := "."

	if builder.source != nil {
		sourcePath = *builder.source
	}

	contextPathSource := fmt.Sprintf("%s/%s", contextPath, sourcePath)

	if _, err := os.Stat(contextPathSource); err != nil {
		if os.IsNotExist(err) {
			return nil, fmt.Errorf("source path %s does not exist", contextPathSource)
		}

		return nil, fmt.Errorf("error checking source path %s: %v", contextPathSource, err)
	}

	// Load root cargo.toml

	sourceCargoPath := fmt.Sprintf("%s/Cargo.toml", contextPathSource)

	if _, err := os.Stat(sourceCargoPath); err != nil {
		if os.IsNotExist(err) {
			return nil, fmt.Errorf("cargo.toml file not found in %s", sourcePath)
		}
		return nil, err
	}

	sourceCargoData, err := os.ReadFile(sourceCargoPath)
	if err != nil {
		return nil, err
	}

	var sourceCargo RustArtifactCargoToml

	_, cargoTomlErr := toml.Decode(string(sourceCargoData), &sourceCargo)
	if cargoTomlErr != nil {
		return nil, cargoTomlErr
	}

	packages := make([]string, 0)
	packagesBinNames := make([]string, 0)
	packagesManifests := make([]string, 0)
	packagesTargets := make([]string, 0)

	if sourceCargo.Workspace != nil && len(sourceCargo.Workspace.Members) > 0 {
		for _, member := range sourceCargo.Workspace.Members {
			packagePath := fmt.Sprintf("%s/%s", contextPathSource, member)
			packageCargoPath := fmt.Sprintf("%s/Cargo.toml", packagePath)

			if _, err := os.Stat(packageCargoPath); err != nil {
				return nil, err
			}

			packageCargoData, err := os.ReadFile(packageCargoPath)
			if err != nil {
				return nil, err
			}

			var packageCargo RustArtifactCargoToml

			_, pkgCargoErr := toml.Decode(string(packageCargoData), &packageCargo)
			if pkgCargoErr != nil {
				return nil, pkgCargoErr
			}

			if len(builder.packages) > 0 && !slices.Contains(builder.packages, packageCargo.Package.Name) {
				continue
			}

			packageTargetPaths := make([]string, 0)

			if packageCargo.Bin != nil && len(packageCargo.Bin) > 0 {
				for _, bin := range packageCargo.Bin {
					packageTargetPath := fmt.Sprintf("%s/%s", packagePath, bin.Path)

					if _, err := os.Stat(packageTargetPath); err != nil {
						return nil, err
					}

					packageTargetPaths = append(packageTargetPaths, packageTargetPath)

					if len(builder.bins) == 0 || slices.Contains(builder.bins, bin.Name) {
						if !slices.Contains(packagesManifests, packageCargoPath) {
							packagesManifests = append(packagesManifests, packageCargoPath)
						}

						packagesBinNames = append(packagesBinNames, bin.Name)
					}
				}
			}

			if len(packageTargetPaths) == 0 {
				packageTargetPath := fmt.Sprintf("%s/src/lib.rs", packagePath)

				if _, err := os.Stat(packageTargetPath); err != nil {
					return nil, err
				}

				packageTargetPaths = append(packageTargetPaths, packageTargetPath)
			}

			for _, memberTargetPath := range packageTargetPaths {
				memberTargetPathRelative := stripPrefix(memberTargetPath, contextPathSource)

				packagesTargets = append(packagesTargets, memberTargetPathRelative)
			}

			packages = append(packages, member)
		}
	}

	// 2. CREATE ARTIFACTS

	// Get rust toolchain artifact

	contextTarget := context.GetTarget()

	rustToolchain, err := artifact.RustToolchain(context)
	if err != nil {
		return nil, err
	}

	rustToolchainTarget, err := artifact.RustToolchainTarget(contextTarget)
	if err != nil {
		return nil, err
	}

	rustToolchainVersion := artifact.RustToolchainVersion()
	rustToolchainName := fmt.Sprintf("%s-%s", rustToolchainVersion, *rustToolchainTarget)

	stepEnvironments := []string{
		"HOME=$VORPAL_WORKSPACE/home",
		fmt.Sprintf(
			"PATH=%s/toolchains/%s/bin",
			artifact.GetEnvKey(rustToolchain),
			rustToolchainName,
		),
		fmt.Sprintf("RUSTUP_HOME=%s", artifact.GetEnvKey(rustToolchain)),
		fmt.Sprintf("RUSTUP_TOOLCHAIN=%s", rustToolchainName),
	}

	// Create vendor artifact

	vendorCargoPaths := []string{"Cargo.toml", "Cargo.lock"}

	for _, workspace := range packages {
		vendorCargoPaths = append(vendorCargoPaths, fmt.Sprintf("%s/Cargo.toml", workspace))
	}

	vendorStepScript, err := template.New("script").Parse(VendorStepScriptTemplate)
	if err != nil {
		return nil, err
	}

	var vendorStepScriptBuffer bytes.Buffer

	stepPackagesTargets := make([]string, 0)
	stepScriptPackages := make([]string, 0)

	for _, pkg := range packagesTargets {
		stepPackagesTargets = append(stepPackagesTargets, fmt.Sprintf("\"%s\"", pkg))
	}

	for _, pkg := range packages {
		stepScriptPackages = append(stepScriptPackages, fmt.Sprintf("\"%s\"", pkg))
	}

	vendorStepScriptArgs := VendorStepScriptTemplateArgs{
		Name:        builder.name,
		Packages:    strings.Join(stepScriptPackages, ","),
		TargetPaths: strings.Join(stepPackagesTargets, " "),
	}

	if err := vendorStepScript.Execute(&vendorStepScriptBuffer, vendorStepScriptArgs); err != nil {
		return nil, err
	}

	stepArtifacts := []*string{rustToolchain}

	vendorStep, err := artifact.Shell(
		context,
		stepArtifacts,
		stepEnvironments,
		vendorStepScriptBuffer.String(),
		builder.secrets,
	)

	vendorName := fmt.Sprintf("%s-vendor", builder.name)

	vendorSource := artifact.NewArtifactSource(vendorName, sourcePath).
		WithIncludes(vendorCargoPaths).
		Build()

	vendorSteps := []*api.ArtifactStep{vendorStep}

	systems := []api.ArtifactSystem{
		api.ArtifactSystem_AARCH64_DARWIN,
		api.ArtifactSystem_AARCH64_LINUX,
		api.ArtifactSystem_X8664_DARWIN,
		api.ArtifactSystem_X8664_LINUX,
	}

	vendorSources := []*api.ArtifactSource{&vendorSource}

	vendor, err := artifact.NewArtifact(vendorName, vendorSteps, systems).
		WithSources(vendorSources).
		Build(context)
	if err != nil {
		return nil, err
	}

	stepArtifacts = append(stepArtifacts, vendor)
	stepArtifacts = append(stepArtifacts, protoc)

	// TODO: implement artifact for 'check` to pre-bake the vendor cache

	sourceIncludes := make([]string, 0)
	sourceExcludes := make([]string, 0)

	sourceExcludes = append(sourceExcludes, "target")

	for _, exclude := range builder.excludes {
		sourceExcludes = append(sourceExcludes, exclude)
	}

	for _, include := range builder.includes {
		sourceIncludes = append(sourceIncludes, include)
	}

	sourceBuilder := artifact.NewArtifactSource(builder.name, sourcePath).
		WithIncludes(sourceIncludes).
		WithExcludes(sourceExcludes)

	source := sourceBuilder.Build()

	for _, artifact := range builder.artifacts {
		stepArtifacts = append(stepArtifacts, artifact)
	}

	sources := []*api.ArtifactSource{&source}

	stepScript, err := template.New("script").Parse(StepScriptTemplate)
	if err != nil {
		return nil, err
	}

	var stepScriptBuffer bytes.Buffer

	if len(packagesBinNames) == 0 {
		packagesBinNames = []string{builder.name}
	}

	if len(packagesManifests) == 0 {
		packagesManifests = []string{sourceCargoPath}
	}

	stepScriptArgs := StepScriptTemplateArgs{
		BinNames:      strings.Join(packagesBinNames, " "),
		Build:         fmt.Sprintf("%t", builder.build),
		Check:         fmt.Sprintf("%t", builder.check),
		Format:        fmt.Sprintf("%t", builder.format),
		Lint:          fmt.Sprintf("%t", builder.lint),
		ManifestPaths: strings.Join(packagesManifests, " "),
		Name:          builder.name,
		Packages:      strings.Join(stepScriptPackages, ","),
		Tests:         fmt.Sprintf("%t", builder.tests),
		Vendor:        artifact.GetEnvKey(vendor),
	}

	if err := stepScript.Execute(&stepScriptBuffer, stepScriptArgs); err != nil {
		return nil, err
	}

	step, err := artifact.Shell(
		context,
		stepArtifacts,
		stepEnvironments,
		stepScriptBuffer.String(),
		builder.secrets,
	)
	if err != nil {
		return nil, err
	}

	steps := []*api.ArtifactStep{step}

	return artifact.NewArtifact(builder.name, steps, systems).
		WithSources(sources).
		Build(context)
}

File: sdk/typescript/src/artifact/language/rust.ts
import { existsSync, readFileSync } from "node:fs";
import { join, relative } from "node:path";
import { parse as parseToml } from "smol-toml";
import type { ArtifactStepSecret } from "../../api/artifact/artifact.js";
import { ArtifactSystem } from "../../api/artifact/artifact.js";
import type { ConfigContext } from "../../context.js";
import {
  ArtifactBuilder,
  ArtifactSourceBuilder,
  getEnvKey,
} from "../../artifact.js";
import { shell } from "../step.js";

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const RUST_TOOLCHAIN_VERSION = "1.89.0";
const PROTOC_ALIAS = "protoc:25.4";

// ---------------------------------------------------------------------------
// Cargo.toml parsing
// ---------------------------------------------------------------------------

/** Represents a `[[bin]]` entry in Cargo.toml. */
interface CargoTomlBinary {
  name: string;
  path: string;
}

/** Represents the `[package]` table in Cargo.toml. */
interface CargoTomlPackage {
  name: string;
}

/** Represents the `[workspace]` table in Cargo.toml. */
interface CargoTomlWorkspace {
  members: string[];
}

/** Minimal representation of a Cargo.toml file. */
interface CargoToml {
  bin: CargoTomlBinary[];
  package?: CargoTomlPackage;
  workspace?: CargoTomlWorkspace;
}

/**
 * Parses a Cargo.toml file using smol-toml.
 *
 * Extracts the fields the Rust builder needs:
 *   - `[package]` name
 *   - `[workspace]` members
 *   - `[[bin]]` name and path
 */
function parseCargo(path: string): CargoToml {
  const content = readFileSync(path, "utf-8");
  const doc = parseToml(content);

  const result: CargoToml = {
    bin: [],
    package: undefined,
    workspace: undefined,
  };

  // Extract [package] name
  if (
    doc.package !== undefined &&
    typeof doc.package === "object" &&
    doc.package !== null &&
    "name" in doc.package &&
    typeof (doc.package as Record<string, unknown>).name === "string"
  ) {
    result.package = {
      name: (doc.package as Record<string, unknown>).name as string,
    };
  }

  // Extract [workspace] members
  if (
    doc.workspace !== undefined &&
    typeof doc.workspace === "object" &&
    doc.workspace !== null &&
    "members" in doc.workspace
  ) {
    const ws = doc.workspace as Record<string, unknown>;
    if (Array.isArray(ws.members)) {
      result.workspace = {
        members: ws.members.filter(
          (m): m is string => typeof m === "string",
        ),
      };
    }
  }

  // Extract [[bin]] entries
  if (Array.isArray(doc.bin)) {
    for (const entry of doc.bin) {
      if (
        typeof entry === "object" &&
        entry !== null &&
        "name" in entry &&
        "path" in entry &&
        typeof (entry as Record<string, unknown>).name === "string" &&
        typeof (entry as Record<string, unknown>).path === "string"
      ) {
        result.bin.push({
          name: (entry as Record<string, unknown>).name as string,
          path: (entry as Record<string, unknown>).path as string,
        });
      }
    }
  }

  return result;
}

// ---------------------------------------------------------------------------
// Rust toolchain target mapping
// ---------------------------------------------------------------------------

/**
 * Maps an ArtifactSystem enum to the Rust target triple.
 * Matches `sdk/go/pkg/artifact/rust_toolchain.go` RustToolchainTarget().
 */
function rustToolchainTarget(system: ArtifactSystem): string {
  switch (system) {
    case ArtifactSystem.AARCH64_DARWIN:
      return "aarch64-apple-darwin";
    case ArtifactSystem.AARCH64_LINUX:
      return "aarch64-unknown-linux-gnu";
    case ArtifactSystem.X8664_DARWIN:
      return "x86_64-apple-darwin";
    case ArtifactSystem.X8664_LINUX:
      return "x86_64-unknown-linux-gnu";
    default:
      throw new Error(`unsupported 'rust-toolchain' system: ${system}`);
  }
}

// ---------------------------------------------------------------------------
// Shell script helpers
// ---------------------------------------------------------------------------

/** Builds the vendor step script for `cargo vendor`. */
function buildVendorScript(
  name: string,
  packages: string[],
  packagesTargets: string[],
): string {
  const lines: string[] = [];

  lines.push(`mkdir -pv $HOME`);
  lines.push(``);
  lines.push(`pushd ./source/${name}-vendor`);

  if (packages.length > 0) {
    const quotedPackages = packages.map((p) => `"${p}"`).join(",");
    const quotedTargets = packagesTargets.map((t) => `"${t}"`).join(" ");

    lines.push(``);
    lines.push(`cat > Cargo.toml << "EOF"`);
    lines.push(`[workspace]`);
    lines.push(`members = [${quotedPackages}]`);
    lines.push(`resolver = "2"`);
    lines.push(`EOF`);

    lines.push(``);
    lines.push(`target_paths=(${quotedTargets})`);
    lines.push(``);
    lines.push(`for target_path in \${target_paths[@]}; do`);
    lines.push(`    mkdir -pv $(dirname \${target_path})`);
    lines.push(`    touch \${target_path}`);
    lines.push(`done`);
  } else {
    lines.push(``);
    lines.push(`mkdir -pv src`);
    lines.push(`touch src/main.rs`);
  }

  lines.push(``);
  lines.push(`mkdir -pv $VORPAL_OUTPUT/vendor`);
  lines.push(``);
  lines.push(
    `cargo_vendor=$(cargo vendor --versioned-dirs $VORPAL_OUTPUT/vendor)`,
  );
  lines.push(``);
  lines.push(`echo "$cargo_vendor" > $VORPAL_OUTPUT/config.toml`);

  return lines.join("\n");
}

/** Builds the main build step script for `cargo build --release`. */
function buildMainScript(opts: {
  name: string;
  vendorDigest: string;
  packages: string[];
  binNames: string[];
  manifests: string[];
  format: boolean;
  lint: boolean;
  check: boolean;
  build: boolean;
  tests: boolean;
}): string {
  const lines: string[] = [];

  lines.push(`mkdir -pv $HOME`);
  lines.push(``);
  lines.push(`pushd ./source/${opts.name}`);
  lines.push(``);
  lines.push(`mkdir -pv .cargo`);
  lines.push(`mkdir -pv $VORPAL_OUTPUT/bin`);
  lines.push(``);
  lines.push(
    `ln -sv ${getEnvKey(opts.vendorDigest)}/config.toml .cargo/config.toml`,
  );

  if (opts.packages.length > 0) {
    const quotedPackages = opts.packages.map((p) => `"${p}"`).join(",");
    lines.push(``);
    lines.push(`cat > Cargo.toml << "EOF"`);
    lines.push(`[workspace]`);
    lines.push(`members = [${quotedPackages}]`);
    lines.push(`resolver = "2"`);
    lines.push(`EOF`);
  }

  lines.push(``);
  lines.push(`bin_names=(${opts.binNames.join(" ")})`);
  lines.push(`manifest_paths=(${opts.manifests.join(" ")})`);

  if (opts.format) {
    lines.push(``);
    lines.push(`echo "Running formatter..."`);
    lines.push(`cargo --offline fmt --all --check`);
  }

  if (opts.lint) {
    lines.push(``);
    lines.push(`for manifest_path in \${manifest_paths[@]}; do`);
    lines.push(`    echo "Running linter..."`);
    lines.push(
      `    cargo --offline clippy --manifest-path \${manifest_path} -- --deny warnings`,
    );
    lines.push(`done`);
  }

  lines.push(``);
  lines.push(`for bin_name in \${bin_names[@]}; do`);

  if (opts.check) {
    lines.push(`    echo "Running check..."`);
    lines.push(`    cargo --offline check --bin \${bin_name} --release`);
  }

  if (opts.build) {
    lines.push(`    echo "Running build..."`);
    lines.push(`    cargo --offline build --bin \${bin_name} --release`);
  }

  if (opts.tests) {
    lines.push(`    echo "Running tests..."`);
    lines.push(`    cargo --offline test --bin \${bin_name} --release`);
  }

  lines.push(`    cp -pv ./target/release/\${bin_name} $VORPAL_OUTPUT/bin/`);
  lines.push(`done`);

  return lines.join("\n");
}

// ---------------------------------------------------------------------------
// RustBuilder
// ---------------------------------------------------------------------------

/**
 * Builder for Rust/Cargo project artifacts.
 *
 * Analogous to:
 * - Rust SDK: `sdk/rust/src/artifact/language/rust.rs` (`Rust` struct)
 * - Go SDK: `sdk/go/pkg/artifact/language/rust.go` (`Rust` struct)
 *
 * The builder:
 * 1. Fetches the rust-toolchain artifact from the registry
 * 2. Fetches the protoc artifact from the registry
 * 3. Parses the Cargo.toml workspace to identify packages, bins, and targets
 * 4. Creates a vendor artifact that runs `cargo vendor`
 * 5. Creates the main build artifact with `cargo build --release`
 *
 * Usage:
 * ```typescript
 * const digest = await new RustBuilder("vorpal", SYSTEMS)
 *   .withPackages(["vorpal-cli"])
 *   .withBins(["vorpal"])
 *   .build(context);
 * ```
 */
export class RustBuilder {
  private _artifacts: string[] = [];
  private _bins: string[] = [];
  private _build: boolean = true;
  private _check: boolean = false;
  private _environments: string[] = [];
  private _excludes: string[] = [];
  private _format: boolean = false;
  private _includes: string[] = [];
  private _lint: boolean = false;
  private _name: string;
  private _packages: string[] = [];
  private _secrets: ArtifactStepSecret[] = [];
  private _source: string | undefined = undefined;
  private _systems: ArtifactSystem[];
  private _tests: boolean = false;

  constructor(name: string, systems: ArtifactSystem[]) {
    this._name = name;
    this._systems = systems;
  }

  /** Adds artifact dependencies available during the build step. */
  withArtifacts(artifacts: string[]): this {
    this._artifacts = artifacts;
    return this;
  }

  /** Sets the binary targets to build. If empty, defaults to the builder name. */
  withBins(bins: string[]): this {
    this._bins = bins;
    return this;
  }

  /** Enables `cargo check` during the build step. */
  withCheck(): this {
    this._check = true;
    return this;
  }

  /** Sets additional environment variables for the build step. Format: "KEY=VALUE". */
  withEnvironments(environments: string[]): this {
    this._environments = environments;
    return this;
  }

  /** Sets file patterns to exclude from the source. */
  withExcludes(excludes: string[]): this {
    this._excludes = excludes;
    return this;
  }

  /** Enables or disables `cargo fmt --check` during the build step. */
  withFormat(format: boolean): this {
    this._format = format;
    return this;
  }

  /** Sets file patterns to include in the source. */
  withIncludes(includes: string[]): this {
    this._includes = includes;
    return this;
  }

  /** Enables or disables `cargo clippy` during the build step. */
  withLint(lint: boolean): this {
    this._lint = lint;
    return this;
  }

  /** Filters which workspace packages to include in the build. */
  withPackages(packages: string[]): this {
    this._packages = packages;
    return this;
  }

  /**
   * Adds secrets available during the build step.
   * Secrets are deduplicated by name.
   */
  withSecrets(secrets: Array<[string, string]>): this {
    for (const [name, value] of secrets) {
      if (!this._secrets.some((s) => s.name === name)) {
        this._secrets.push({ name, value });
      }
    }
    return this;
  }

  /** Sets the source path relative to the artifact context directory. */
  withSource(source: string): this {
    this._source = source;
    return this;
  }

  /** Enables or disables `cargo test` during the build step. */
  withTests(tests: boolean): this {
    this._tests = tests;
    return this;
  }

  /**
   * Builds the Rust project artifact.
   *
   * @returns The artifact digest string
   */
  async build(context: ConfigContext): Promise<string> {
    // Sort secrets for deterministic output
    this._secrets.sort((a, b) => a.name.localeCompare(b.name));

    // Fetch protoc artifact
    const protoc = await context.fetchArtifactAlias(PROTOC_ALIAS);

    // Parse source path
    const contextPath = context.getArtifactContextPath();
    const sourcePath = this._source ?? ".";
    const contextPathSource = join(contextPath, sourcePath);

    if (!existsSync(contextPathSource)) {
      throw new Error(
        `\`source.${this._name}.path\` not found: ${sourcePath}`,
      );
    }

    // Parse Cargo.toml
    const sourceCargoPath = join(contextPathSource, "Cargo.toml");

    if (!existsSync(sourceCargoPath)) {
      throw new Error(`Cargo.toml not found: ${sourceCargoPath}`);
    }

    const sourceCargo = parseCargo(sourceCargoPath);

    // Get list of bin targets
    const packages: string[] = [];
    const packagesBinNames: string[] = [];
    const packagesManifests: string[] = [];
    const packagesTargets: string[] = [];

    if (
      sourceCargo.workspace !== undefined &&
      sourceCargo.workspace.members.length > 0
    ) {
      for (const member of sourceCargo.workspace.members) {
        const packagePath = join(contextPathSource, member);
        const packageCargoPath = join(packagePath, "Cargo.toml");

        if (!existsSync(packageCargoPath)) {
          throw new Error(`Cargo.toml not found: ${packageCargoPath}`);
        }

        const packageCargo = parseCargo(packageCargoPath);

        if (
          this._packages.length > 0 &&
          packageCargo.package !== undefined &&
          !this._packages.includes(packageCargo.package.name)
        ) {
          continue;
        }

        const packageTargetPaths: string[] = [];

        if (packageCargo.bin.length > 0) {
          for (const bin of packageCargo.bin) {
            const packageTargetPath = join(packagePath, bin.path);

            if (!existsSync(packageTargetPath)) {
              throw new Error(
                `bin target not found: ${packageTargetPath}`,
              );
            }

            packageTargetPaths.push(packageTargetPath);

            if (
              this._bins.length === 0 ||
              this._bins.includes(bin.name)
            ) {
              if (!packagesManifests.includes(packageCargoPath)) {
                packagesManifests.push(packageCargoPath);
              }

              packagesBinNames.push(bin.name);
            }
          }
        }

        if (packageTargetPaths.length === 0) {
          const packageTargetPath = join(packagePath, "src/lib.rs");

          if (!existsSync(packageTargetPath)) {
            throw new Error(
              `lib.rs not found: ${packageTargetPath}`,
            );
          }

          packageTargetPaths.push(packageTargetPath);
        }

        for (const memberTargetPath of packageTargetPaths) {
          const memberTargetPathRelative = relative(
            contextPathSource,
            memberTargetPath,
          );
          packagesTargets.push(memberTargetPathRelative);
        }

        packages.push(member);
      }
    }

    // 2. CREATE ARTIFACTS

    // Get rust toolchain artifact
    const rustToolchain = await context.fetchArtifactAlias(
      `rust-toolchain:${RUST_TOOLCHAIN_VERSION}`,
    );

    const rustToolchainTargetStr = rustToolchainTarget(context.getSystem());
    const rustToolchainName = `${RUST_TOOLCHAIN_VERSION}-${rustToolchainTargetStr}`;

    const stepEnvironments: string[] = [
      "HOME=$VORPAL_WORKSPACE/home",
      `PATH=${getEnvKey(rustToolchain)}/toolchains/${rustToolchainName}/bin`,
      `RUSTUP_HOME=${getEnvKey(rustToolchain)}`,
      `RUSTUP_TOOLCHAIN=${rustToolchainName}`,
    ];

    for (const env of this._environments) {
      stepEnvironments.push(env);
    }

    // Create vendor artifact

    const vendorCargoPaths: string[] = ["Cargo.toml", "Cargo.lock"];

    for (const pkg of packages) {
      vendorCargoPaths.push(`${pkg}/Cargo.toml`);
    }

    const vendorStepScript = buildVendorScript(
      this._name,
      packages,
      packagesTargets,
    );

    let stepArtifacts: string[] = [rustToolchain];

    const vendorStep = await shell(
      context,
      stepArtifacts,
      stepEnvironments,
      vendorStepScript,
      this._secrets,
    );

    const vendorName = `${this._name}-vendor`;

    const vendorSource = new ArtifactSourceBuilder(vendorName, sourcePath)
      .withIncludes(vendorCargoPaths)
      .build();

    const vendor = await new ArtifactBuilder(
      vendorName,
      [vendorStep],
      this._systems,
    )
      .withSources([vendorSource])
      .build(context);

    stepArtifacts = [...stepArtifacts, vendor, protoc];

    // Create source

    const sourceIncludes: string[] = [];
    const sourceExcludes: string[] = ["target"];

    for (const exclude of this._excludes) {
      sourceExcludes.push(exclude);
    }

    for (const include of this._includes) {
      sourceIncludes.push(include);
    }

    const sourceBuilder = new ArtifactSourceBuilder(this._name, sourcePath)
      .withIncludes(sourceIncludes)
      .withExcludes(sourceExcludes);

    const source = sourceBuilder.build();

    for (const artifact of this._artifacts) {
      stepArtifacts.push(artifact);
    }

    // Create step

    if (packagesBinNames.length === 0) {
      packagesBinNames.push(this._name);
    }

    if (packagesManifests.length === 0) {
      packagesManifests.push(sourceCargoPath);
    }

    const stepScript = buildMainScript({
      name: this._name,
      vendorDigest: vendor,
      packages,
      binNames: packagesBinNames,
      manifests: packagesManifests,
      format: this._format,
      lint: this._lint,
      check: this._check,
      build: this._build,
      tests: this._tests,
    });

    const step = await shell(
      context,
      stepArtifacts,
      stepEnvironments,
      stepScript,
      this._secrets,
    );

    // Create artifact

    return new ArtifactBuilder(this._name, [step], this._systems)
      .withSources([source])
      .build(context);
  }
}
